provider/azure.go

```diff
@@ -0,0 +1,456 @@
+	if err != nil {
+		return nil, err
+	}
+
+	for _, zone := range zones {
+		err := p.iterateRecords(*zone.Name, func(recordSet dns.RecordSet) bool {
+			if recordSet.Name == nil || recordSet.Type == nil {
+				log.Error("Skipping invalid record set with nil name or type.")
+				return true
+			}
+			recordType := strings.TrimLeft(*recordSet.Type, "Microsoft.Network/dnszones/")
+			switch dns.RecordType(recordType) {
+			case dns.A, dns.CNAME, dns.TXT:
+				name := formatDNSName(*recordSet.Name, *zone.Name)
+				target := extractTarget(recordType, &recordSet)
+				if target == "" {
+					log.Errorf("Failed to extract target for '%s' with type '%s'.", name, recordType)
+					return true
+				}
+				endpoint := endpoint.NewEndpoint(name, target, recordType)
+				log.Debugf(
+					"Adding %s record for '%s' with target '%s'.",
+					endpoint.RecordType,
+					endpoint.DNSName,
+					endpoint.Target,
+				)
+				endpoints = append(endpoints, endpoint)
+			default:
+			}
+			return true
+		})
+		if err != nil {
+			return nil, err
+		}
+	}
+	return endpoints, nil
+}
+
+// ApplyChanges applies the given changes.
+//
+// Returns nil if the operation was successful or an error if the operation failed.
+func (p *AzureProvider) ApplyChanges(changes *plan.Changes) error {
+	zones, err := p.azureZones()
+	if err != nil {
+		return err
+	}
+
+	deleted, updated := mapAzureChanges(zones, changes)
+	p.deleteAzureRecords(deleted)
+	p.updateAzureRecords(updated)
+	return nil
+}
+
+func (p *AzureProvider) azureZones() ([]dns.Zone, error) {
+	log.Debug("Retrieving Azure DNS zones.")
+
+	var zones []dns.Zone
+	list, err := p.zonesClient.ListByResourceGroup(p.resourceGroup, nil)
+	if err != nil {
+		return zones, err
+	}
+
+	for list.Value != nil && len(*list.Value) > 0 {
+		for _, zone := range *list.Value {
+			if zone.Name != nil && strings.HasSuffix(*zone.Name, p.domainFilter) {
+				zones = append(zones, zone)
+			}
+		}
+
+		list, err = p.zonesClient.ListNextResults(list)
+		if err != nil {
+			return zones, err
+		}
+	}
+	log.Debugf("Found %d Azure DNS zone(s).", len(zones))
+	return zones, nil
+}
+
+func (p *AzureProvider) iterateRecords(zoneName string, callback func(dns.RecordSet) bool) error {
+	log.Debugf("Retrieving Azure DNS records for zone '%s'.", zoneName)
+
+	list, err := p.recordsClient.ListByDNSZone(p.resourceGroup, zoneName, nil)
+	if err != nil {
+		return err
+	}
+
+	done := false
+	for list.Value != nil && len(*list.Value) > 0 {
+		for _, recordSet := range *list.Value {
+			if !callback(recordSet) {
+				done = true
+				break
+			}
+		}
+
+		if done {
+			break
+		}
+
+		list, err = p.recordsClient.ListByDNSZoneNextResults(list)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func formatDNSName(recordName, zoneName string) string {
+	if recordName == "@" {
+		return zoneName
+	}
+	return fmt.Sprintf("%s.%s", recordName, zoneName)
+}
+
+func extractTarget(recordType string, recordSet *dns.RecordSet) string {
```

Comment:
I don't believe the properties can contain data for multiple types (this structure is effectively a union afaik); it that's true, we don't need to pass the type.  Let me look into that.

Regarding `dns.RecordSet.Type`, we're not using it here since we already stripped off the `Microsoft.Network/dnszones/` prefix above and I'd like the code not to have to take it into account again.